aiofreepybox/aiofreepybox.py

```diff
@@ -1,114 +1,285 @@
+                    __name__.rsplit(".", 1)[0] + ".api", str(n) + ".py"
+                ),
+                api_l,
+            )
+        )
+        if len(api_l) != l:
+            raise KeyError
+
+        mods = {
+            str(n): import_module(".api." + str(n), package=__name__.rsplit(".", 1)[0])
+            for n in api_l
+        }
+        return mods
+
+    def __getattr__(self, k):
+        """ Return API attribute """
+
+        if "__" in k:
+            raise AttributeError
+
+        try:
+            mod = self.__imp_api__([k])
+        except KeyError:
+            _LOGGER.error(f"Invalid API name: {k}")
+            raise AttributeError
+        else:
+            if self._access:
+                kwargs: Dict[str, str] = {}
+                if k in _API_MODS_PARAMS and isinstance(_API_MODS_PARAMS[k], dict):
+                    kwargs = _API_MODS_PARAMS[k]
+                setattr(
+                    self, k, getattr(mod[k], k.capitalize())(self._access, **kwargs)
+                )
+            else:
+                return getattr(mod[k], k.capitalize())
+
+        return getattr(self, k)
+
+    def clean_db(self, uids: Optional[list] = None, all_: bool = False) -> bool:
+        """
+        Remove unauth fbx's from db
+
+        uids : `list`, optional
+            , Default to None
+        all_ : `bool`
+            , Default to False
+        """
+
+        if uids is None:
+            uids = [
+                base64.b64decode(x.name.rsplit("_", 1)[1]).decode("utf-8")
+                for x in Path(self.data_dir).glob(_F_DB_NAME + "_*")
+            ]
+        auth = [
+            base64.b64decode(x.name.rsplit("_", 1)[1]).decode("utf-8")
+            for x in Path(self.data_dir).glob(_F_TOKEN_NAME + "_*")
+        ]
+        c = 0
+        for uid in uids:
+            if uid not in auth or all_:
+                fname = Path(self.data_dir).joinpath(
+                    _F_DB_NAME
+                    + "_"
+                    + base64.b64encode(uid.encode("utf-8")).decode("utf-8")
+                )
+                try:
+                    fname.unlink()
+                    c += 1
+                except FileNotFoundError:
+                    pass
+        if c != 0:
+            _LOGGER.debug(f"{c} db entry cleared")
+            return True
+        return False
+
+    async def close(self) -> None:
+        """
         Close the freebox session
-        '''
-        if self._access is None:
-            raise NotOpenError('Freebox is not open')
+        """
+
+        if self._session and self._session.closed:  # type: ignore # noqa
+            return None
+
+        try:
+            await self._access.post("login/logout")  # type: ignore # noqa
+        except AttributeError:
+            pass
+        try:
+            await self._session.close()  # type: ignore # noqa
+            await asyncio.sleep(0.250)
+        except AttributeError:
+            pass
+
+    async def discover(
```

Comment:
Wow, we went from a 50-lines auto-discovery to something that is around 400/500 lines, what happened?! The first version seemed to work, was it incorrect? I tried understanding the new stuff you did with the various disc_* functions, the db, etc., but it's too complex for me, I can't get my head around it... I'll come back to it with a fresher eye, but in the meantime I reiterate my proposal to simply transfer the project ownership to you if you want to move forward without waiting for my approval.

Reply:
No worries, I'll explain in a few words (or maybe a bit more), it all started when @SNoof85 opened issue #52.
The first version was ok, but it had it's own limits, like the one @SNoof85 raised. Technically this issue (#52) had been here long before, but it was up to the user to figure out a way to workaround it (which was both kind of painful and complicated). This is the result, and it came with lots of perks, first ones being, the uid and the db, now you can manage a whole fleet of freeboxes easily, no more switching token files and making sure you've set the correct ip adress, now once setup is done you're good to go.
As i was writing it, it turned out to be complex and unreadable.
Here is a quick description of what it does so you can get around it quickly:

discover:
It still tries to find a freebox on the lan after checking it has not opened a session already, then updates the db if it found something.

open:
Now, first it will try to find the uid of the freebox you're requesting if it was not provided (uid is now the preferred way to adress any freebox) but it still supports the "host" "port" format,  otherwise it will load the db for the specified uid.
Then as before it will try to open a session, but unlike before, it will try every way it knows to contact the freebox until it fails if the freebox can't be reached.

Each freebox creates two files, the app_auth (now named ".fbx_app_auth_{uid_base64}" ) like before and the db (named " .fbx_db_{uid_base64}") those can now be stored wherever you want using the "data_dir" option, they can be read with bzcat.
The db store known freebox adresses, linked with the fbx_desc. It can be managed with get_db and clean_db.

The rest is made up of support functions for discover and open as they can combine in some  ways and db management.

Reply:
Thanks for the thorough explanation!